src/libappimage/core/PayloadIterator.cpp

```diff
@@ -0,0 +1,189 @@
+            AppImage appImage;
+
+            // to be used by the read method when the end of the traversal is reached
+            std::stringstream emptyStream;
+
+            // Real Traversal implementation
+            std::shared_ptr<Traversal> traversal;
+
+            // flags whether a the current entry contents has been read or not
+            bool entryDataConsumed = false;
+        public:
+            /**
+             * Initialized the Private class with required traversal derivative if <atEnd> is false.
+             * @param appImage
+             * @param atEnd determine if a end state iterator should be created
+             */
+            explicit Private(const AppImage& appImage, bool atEnd = false);
+
+            // Creating copies of this object is not allowed
+            Private(Private& other) = delete;
+
+            // Creating copies of this object is not allowed
+            Private& operator=(Private& other) = delete;
+
+            // Move constructor
+            Private(Private&& other) noexcept;
+
+            // Move assignment operator
+            Private& operator=(Private&& other) noexcept;
+
+            /**
+             * Compare Private data according to the AppImage they point to and to the traversal instance.
+             * @param rhs
+             * @return
+             */
+            bool operator==(const Private& rhs) const;
+
+            bool operator!=(const Private& rhs) const;
+
+            bool isCompleted() { return traversal == nullptr; }
+
+            void next();
+
+            PayloadEntryType type() { return isCompleted() ? PayloadEntryType::UNKNOWN : traversal->getEntryType(); }
+
+            std::string entryName() { return isCompleted() ? std::string() : traversal->getEntryName(); }
+
+            std::string entryLink() { return isCompleted() ? std::string() : traversal->getEntryLink(); }
+
+            void extractTo(const std::string& target) {
+                // Enforce ONE PASS restriction
+                if (entryDataConsumed)
+                    throw PayloadIteratorError("Entry data consumed");
+                else
+                    entryDataConsumed = true;
+
+                if (!isCompleted()) traversal->extract(target);
+            }
+
+            std::istream& read() {
+                // Enforce ONE PASS restriction
+                if (entryDataConsumed)
+                    throw PayloadIteratorError("Entry data consumed");
+                else
+                    entryDataConsumed = true;
+
+                return isCompleted() ? emptyStream : traversal->read();
+            }
+
+            Private* beginState() { return new Private(appImage); }
+
+            Private* endState() { return new Private(appImage, true); }
+        };
+
+        PayloadIterator::PayloadIterator(const AppImage& appImage) : d(new Private(appImage)) {}
+
+        PayloadIterator::PayloadIterator(PayloadIterator&& other) noexcept { d = other.d; }
```

Comment:
Please reformat the file not to have any expressions in the method bodies on the same line as the signature definition. That'll make the file consistent with the rest of the codebase style wise.